include/cantera/kinetics/MultiRate.h

```diff
@@ -0,0 +1,115 @@
+/**
+ * @file MultiRate.h
+ *
+ * @warning This file is an experimental part of the %Cantera API and
+ *     may be changed or removed without notice.
+ */
+
+// This file is part of Cantera. See License.txt in the top-level directory or
+// at https://cantera.org/license.txt for license and copyright information.
+
+#ifndef CT_MULTIRATE_H
+#define CT_MULTIRATE_H
+
+#include "cantera/kinetics/ReactionRate.h"
+
+namespace Cantera
+{
+
+//! An abstract base class for evaluating all reactions of a particular type.
+/**
+ * Because this class has no template parameters, the `Kinetics` object
+ * can store all of these rate coefficient evaluators as a
+ * `vector<shared_ptr<MultiRateBase>>`.
+ *
+ * @todo At the moment, implemented methods are specific to `BulkKinetics`,
+ *     which can be updated using information of a single `ThermoPhase`.
+ *     `InterfaceKinetics` will require access to an entire `Kinetics` object
+ *     or the underlying `vector<ThermoPhase*>` vector (e.g. `m_thermo`).
+ *
+ * @warning This class is an experimental part of the %Cantera API and
+ *     may be changed or removed without notice.
+ */
+class MultiRateBase
+{
+public:
+    virtual ~MultiRateBase() {}
+
+    //! Add reaction rate object to the evaluator
+    virtual void add(const size_t rxn_index,
+                     const shared_ptr<ReactionRateBase> rate) = 0;
+
+    //! Replace reaction rate object handled by the evaluator
+    virtual bool replace(const size_t rxn_index,
+                         const shared_ptr<ReactionRateBase> rate) = 0;
+
+    //! Evaluate all rate constants handled by the evaluator
+    virtual void getRateConstants(const ThermoPhase& bulk_phase,
+                                  double* kf) const = 0;
+
+    //! Update data common to reaction rates of a specific type
+    virtual void update(const ThermoPhase& bulk) = 0;
+};
+
+
+//! A class template handling all reaction rates specific to `BulkKinetics`.
+/**
+ * @warning This class is an experimental part of the %Cantera API and
+ *     may be changed or removed without notice.
+ */
+template <class RateType, class DataType>
+class MultiBulkRates final : public MultiRateBase
+{
+public:
+    virtual void add(const size_t rxn_index,
+                     const shared_ptr<ReactionRateBase> rate) override {
+        m_indices[rxn_index] = m_rates.size();
+        m_rates.push_back(*std::dynamic_pointer_cast<RateType>(rate));
```

Comment:
Since this class is just making a copy of the `RateType` object, there's not much benefit to having it take a `shared_ptr` argument. The following modification would also cause a type mismatch to raise an exception rather than resulting in a null pointer dereference as is currently the case:
```suggestion
    virtual void add(const size_t rxn_index,
                     const ReactionRateBase* rate) override {
        m_indices[rxn_index] = m_rates.size();
        m_rates.push_back(dynamic_cast<RateType&>(*rate));
```

Reply:
There indeed isn't much of a point in the shared pointer. Had to tweak that a little bit (`dynamic_cast` causes issues for `const`), and passing the actual object by reference works just fine.
```C++
    virtual void add(const size_t rxn_index,
                     ReactionRateBase& rate) override {
        m_indices[rxn_index] = m_rates.size();
        m_rates.push_back(dynamic_cast<RateType&>(rate));
        m_rxn.push_back(rxn_index);
    }
```